sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1713,8 +1713,7 @@ object PushPredicateThroughNonJoin extends Rule[LogicalPlan] with PredicateHelpe
       project.copy(child = Filter(replaceAlias(condition, aliasMap), grandChild))
 
     case filter @ Filter(condition, aggregate: Aggregate)
-      if aggregate.aggregateExpressions.forall(_.deterministic)
```

Comment:
Do you investigate the issue https://issues.apache.org/jira/browse/SPARK-13473.
It seems we can't release the restrictions.

Reply:
I investigated before, 
Similar ones are https://issues.apache.org/jira/browse/SPARK-20246. 
At that time, it was because it was impossible to determine whether the expression was deterministic for example `$"rand" > 5`, therefore, it will incorrectly push down `$"rand" > 5`. 
However, after the filter expression is replaced, it can be judged whether it is a deterministic expression.

Reply:
SGTM.

Reply:
If so, how about keep
`if aggregate.aggregateExpressions.exists(_.deterministic)`

Reply:
`if aggregate.groupingExpressions.nonEmpty`, does it necessarily satisfy `aggregate.aggregateExpressions.exists(_.deterministic)`?

Reply:
Updated.